Modify interaction developer experience improvements

ahocevar opened the pull request:
Working with the Modify interaction in an application, I noticed several flaws regarding developer experience:

* When modifying points with a style that is different from a circle with 10px radius, it is hard to grab the symbol for dragging because the vertex selection is geometry based. I would expect being able to grab the point for dragging anywhere (like it can currently be done with the Translate interaction). This could easily be achieved by adding hit detection.
* When using the `modifystart` and `modifyend` events, the `features` in the event object are all features on the modify `source`. But I would expect the features to be only those actually being modified.
* When configuring a `style` for the interaction, the vertex feature has no indication about which feature(s) it belongs to. So it is impossible to do conditional styling (e.g. only render a drag vertex for lines, not for points). I'd expect the feature to have a `features` property, so I can apply conditional styling based on the features the vertex belongs to.

This pull request fixes these flaws.

To take advantage of hit detection for points, the `Modify` interaction has to be configured with the new `layer` option.

Fixes #7404 
Replaces #8871
Fixes #8242 (thanks to the added hit detection, modifystart/modifyend can now be used)

ahocevar commented:
This pull request is now ready for review.

ahocevar commented:
I added another commit that simplifies both code and API for hit detection mode, and an example to show the new `layer` option. This should make it easier to review.

MoonE reviewed (commented):
-If I set a negative y-scale on the icon image the hit still works on the unmodified icon.-
I did not properly change the scale, it works fine...

ahocevar commented on src/ol/interaction/Modify.js:
```diff
@@ -314,27 +332,44 @@ class Modify extends PointerInteraction {
+          VectorEventType.REMOVEFEATURE,
+          this.handleSourceRemove_.bind(this)
+        );
+      }
```
Thanks for raising this question. It made me re-think the API, and I decided to abandon the new `layer` option. Instead, I added a new `hitDetection` option, which can be configured with a boolean (for detection of the feature on any layer) or a layer (for detection of the feature on a specific layer).

ahocevar commented:
Thanks for the review, @MoonE. I have addressed your comments.

MoonE commented on src/ol/interaction/Modify.js:
```diff
@@ -376,14 +408,27 @@ class Modify extends PointerInteraction {
+  willModifyFeatures_(evt, segments) {
+    if (!this.featuresBeingModified_) {
+      this.featuresBeingModified_ = new Collection();
+      const features = this.featuresBeingModified_.getArray();
+      for (let i = 0, ii = segments.length; i < ii; ++i) {
+        const feature = segments[i][0].feature;
+        if (features.indexOf(feature) === -1) {
+          this.featuresBeingModified_.push(feature);
+        }
+      }
+
       this.dispatchEvent(
-        new ModifyEvent(ModifyEventType.MODIFYSTART, this.features_, evt)
+        new ModifyEvent(
+          ModifyEventType.MODIFYSTART,
+          new Collection(features),
```
This should be `this.featuresBeingModified_` instead of a new Collection.

ahocevar replied:
Thanks, fixed.

MoonE commented on src/ol/interaction/Modify.js:
```diff
@@ -108,14 +108,24 @@ const ModifyEventType = {
- * must be provided with the features option.
+ * must be provided with the `features` option.
+ * @property {boolean|import("../layer/BaseVector").default} [hitDetection] When configured, point
+ * features will considered for modification based on their visual appearance, instead of being within
```
```suggestion
 * features will be considered for modification based on their visual appearance, instead of being within
```

MoonE reviewed (changes requested):
If hit detection is enabled with a point and line feature and the hit is detected on the point feature a point is inserted on the line if that point is closer to the clicked position.
I updated the new example here to show the issue. If you grab the icon at the top it will instead insert a new vertex into the line.
https://github.com/MoonE/openlayers/tree/modify-dx-bug

ahocevar commented:
Thanks for testing this, @MoonE. I have addressed your comment - in your example, no line vertex won't be inserted any more when the icon is hit detected.

MoonE commented:
Thanks you, @ahocevar, unfortunately I found another bug. I updated the example in https://github.com/MoonE/openlayers/tree/modify-dx-bug.

If you drag the icon it moves both of the  green line at [0, 0] it should only move one.
If you then alt click the point feature to delete, it only deletes one of the points for the green line, but seems to delete both from its internal rbush. The internal representation is out of sync from then on.

I also updated the code for the `pointermove` event in the example to better work with hitdetection and polygons. It should only change the cursor when it is near one of the lines not when it is anywhere over the polygon. It is probably better to use this in the example because this should always work.

ahocevar commented:
@MoonE I just rebased this pull request to bring in your changes from #11803. So the other bug you noticed should be fixed now?

I also added a commit, taking inspiration from your change to get a besser cursor feedback when hovering over modification candidates, removing the overhead of another hit detection cycle.

MoonE commented:
Yes, this seems to work now.

Something else I noticed:
If you modify a point feature it sometimes gets grabbed and dragged wherever it was clicked and sometimes it is placed directly under the pointer. This depends on the distance of the click from the point coordinate
Maybe it should always use the new behavior to not center the vertex under the cursor? It feels better to me and would be more consistent.

ahocevar commented:
Good suggestion, @MoonE. Changed with 3e3cd08.

ahocevar commented:
For the previous change, I had to increase a few test tolerances where circle radiuses are calculated from pointer locations, because now the pointer location counts, instead of the vertex geometry. Ready for review again.

MoonE commented on test/spec/ol/interaction/modify.test.js:
```diff
@@ -1000,7 +1084,7 @@ describe('ol.interaction.Modify', function () {
       simulateEvent('pointerup', 30, -5, null, 0);
 
-      expect(circleFeature.getGeometry().getRadius()).to.equal(25);
+      expect(circleFeature.getGeometry().getRadius()).to.roughlyEqual(25, 1e9);
```
The tolerance is way too high here, either should be 1e-9 or 0.1 like the others.

ahocevar replied:
Right, thanks for catching this! Fixed.

MoonE reviewed (approved):
Great work, @ahocevar. Everything looks good now.

ahocevar commented:
Thanks a lot @MoonE for your sharp eye and all the help on this!
